File: go/parsers/invocation.go
package parsers

import (
	"google.golang.org/protobuf/proto"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-go/serializers"
	"github.com/xconnio/wampproto-protobuf/go/gen"
)

type Invocation struct {
	gen *gen.Invocation
	ex  *PayloadExpander
}

func NewInvocationFields(gen *gen.Invocation, payload []byte) messages.InvocationFields {
	return &Invocation{
		gen: gen,
		ex:  &PayloadExpander{payload: payload, serializer: gen.GetPayloadSerializerId()},
	}
}

func (i *Invocation) RequestID() uint64 {
	return i.gen.GetRequestId()
}

func (i *Invocation) RegistrationID() uint64 {
	return i.gen.GetRegistrationId()
}

func (i *Invocation) Details() map[string]any {
	var details map[string]any

	if i.gen.Caller > 0 {
		setDetail(&details, "caller", i.gen.Caller)
	}

	if i.gen.CallerAuthid != "" {
		setDetail(&details, "caller_authid", i.gen.CallerAuthid)
	}

	if i.gen.CallerAuthrole != "" {
		setDetail(&details, "caller_authrole", i.gen.CallerAuthrole)
	}

	if i.gen.Procedure != "" {
		setDetail(&details, "procedure", i.gen.Procedure)
	}

	return details
}

func (i *Invocation) Args() []any {
	return i.ex.Args()
}

func (i *Invocation) KwArgs() map[string]any {
	return i.ex.Kwargs()
}

func (i *Invocation) PayloadIsBinary() bool {
	return true
}

func (i *Invocation) Payload() []byte {
	return i.ex.Payload()
}

func (i *Invocation) PayloadSerializer() uint64 {
	return i.gen.GetPayloadSerializerId()
}

func InvocationToProtobuf(invocation *messages.Invocation) ([]byte, error) {
	msg := &gen.Invocation{
		RequestId:      invocation.RequestID(),
		RegistrationId: invocation.RegistrationID(),
	}

	payloadSerializer := selectPayloadSerializer(invocation.Details())
	msg.PayloadSerializerId = payloadSerializer

	var payload []byte
	if invocation.PayloadIsBinary() {
		payload = invocation.Payload()
	} else {
		var err error
		payload, err = serializers.SerializePayload(payloadSerializer, invocation.Args(), invocation.KwArgs())
		if err != nil {
			return nil, err
		}
	}

	data, err := proto.Marshal(msg)
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeInvocation, data, payload), nil
}

func ProtobufToInvocation(data, payload []byte) (*messages.Invocation, error) {
	msg := &gen.Invocation{}
	err := proto.Unmarshal(data, msg)
	if err != nil {
		return nil, err
	}

	return messages.NewInvocationWithFields(NewInvocationFields(msg, payload)), nil
}

File: python/wampprotobuf/parsers/invocation.py
from typing import Any

from wampproto.messages import Message
from wampproto.messages.invocation import IInvocationFields, Invocation

from wampprotobuf.gen import invocation_pb2
from wampprotobuf.parsers import helpers


class InvocationFields(IInvocationFields):
    def __init__(self, msg: invocation_pb2.Invocation):
        super().__init__()
        self._msg = msg
        self._args = None
        self._kwargs = None
        self._unpacked = False

    @property
    def request_id(self) -> int:
        return self._msg.request_id

    @property
    def registration_id(self) -> int:
        return self._msg.registration_id

    def unpack(self):
        try:
            self._unpacked = True
            args, kwargs = helpers.from_cbor_payload(self._msg.payload)
            self._args = args
            self._kwargs = kwargs
        except Exception as e:
            print(f"error parsing CBOR payload: {e}")

    @property
    def args(self) -> list[Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._args

    @property
    def kwargs(self) -> dict[str, Any] | None:
        if not self._unpacked:
            self.unpack()

        return self._kwargs

    @property
    def details(self) -> dict[str, Any]:
        return {}

    def payload_is_binary(self) -> bool:
        return True

    @property
    def payload(self) -> bytes | None:
        return self._msg.payload

    @property
    def payload_serializer(self) -> int:
        return self._msg.payload_serializer


def from_protobuf(payload: bytes) -> Message:
    result = invocation_pb2.Invocation()
    result.ParseFromString(payload)

    return Invocation(InvocationFields(result))


def to_protobuf(invocation: Invocation) -> bytes:
    result = invocation_pb2.Invocation()
    result.request_id = invocation.request_id
    result.registration_id = invocation.registration_id

    if invocation.payload_is_binary():
        result.payload = invocation.payload
        result.payload_serializer = invocation.payload_serializer
    else:
        try:
            payload = helpers.to_cbor_payload(invocation.args, invocation.kwargs)
        except Exception as e:
            raise Exception(f"error in serialization to cbor: {e}")

        serializer = 1

        result.payload = payload
        result.payload_serializer = serializer

    return result.SerializeToString()
